File: server/email-service.ts
// Resend integration for sending contact form emails
import { Resend } from 'resend';

let connectionSettings: any;

async function getCredentials() {
  const hostname = process.env.REPLIT_CONNECTORS_HOSTNAME;
  const xReplitToken = process.env.REPL_IDENTITY 
    ? 'repl ' + process.env.REPL_IDENTITY 
    : process.env.WEB_REPL_RENEWAL 
    ? 'depl ' + process.env.WEB_REPL_RENEWAL 
    : null;

  if (!xReplitToken) {
    throw new Error('X_REPLIT_TOKEN not found for repl/depl');
  }

  connectionSettings = await fetch(
    'https://' + hostname + '/api/v2/connection?include_secrets=true&connector_names=resend',
    {
      headers: {
        'Accept': 'application/json',
        'X_REPLIT_TOKEN': xReplitToken
      }
    }
  ).then(res => res.json()).then(data => data.items?.[0]);

  if (!connectionSettings || (!connectionSettings.settings.api_key)) {
    throw new Error('Resend not connected');
  }
  return { apiKey: connectionSettings.settings.api_key, fromEmail: connectionSettings.settings.from_email };
}

async function getUncachableResendClient() {
  const { apiKey, fromEmail } = await getCredentials();
  return {
    client: new Resend(apiKey),
    fromEmail
  };
}

export interface ContactFormData {
  name: string;
  email: string;
  message: string;
}

// Escape HTML to prevent injection attacks
function escapeHtml(text: string): string {
  const htmlEntities: Record<string, string> = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#39;'
  };
  return text.replace(/[&<>"']/g, char => htmlEntities[char]);
}

// Recipient email - configure via environment variable for production
const CONTACT_RECIPIENT = process.env.CONTACT_EMAIL || 'morthalasharath@gmail.com';

export async function sendContactEmail(data: ContactFormData): Promise<{ success: boolean; error?: string }> {
  try {
    const { client } = await getUncachableResendClient();
    
    // Sanitize all user input before inserting into HTML
    const safeName = escapeHtml(data.name);
    const safeEmail = escapeHtml(data.email);
    const safeMessage = escapeHtml(data.message);
    
    // Use Resend's test sender - this works for sending to your own verified email
    // To use a custom domain, verify it at https://resend.com/domains
    const result = await client.emails.send({
      from: 'Backend Studio <onboarding@resend.dev>',
      to: CONTACT_RECIPIENT,
      subject: `Contact Form: Message from ${safeName}`,
      html: `
        <h2>New Contact Form Submission</h2>
        <p><strong>From:</strong> ${safeName}</p>
        <p><strong>Email:</strong> ${safeEmail}</p>
        <hr />
        <h3>Message:</h3>
        <p>${safeMessage.replace(/\n/g, '<br>')}</p>
        <hr />
        <p style="color: #666; font-size: 12px;">
          This message was sent from the Backend Systems Intelligence Studio contact form.
        </p>
      `,
      text: `
New Contact Form Submission

From: ${data.name}
Email: ${data.email}

Message:
${data.message}

---
This message was sent from the Backend Systems Intelligence Studio contact form.
      `
    });

    if (result.error) {
      console.error('Resend error:', result.error);
      return { success: false, error: result.error.message };
    }

    return { success: true };
  } catch (error) {
    console.error('Email service error:', error);
    return { success: false, error: error instanceof Error ? error.message : 'Failed to send email' };
  }
}

File: api/email_service.py
import os
import httpx
import html
from typing import Optional
import resend

CONTACT_RECIPIENT = os.environ.get("CONTACT_EMAIL", "morthalasharath@gmail.com")

def escape_html(text: str) -> str:
    return html.escape(text)

async def get_resend_credentials():
    hostname = os.environ.get("REPLIT_CONNECTORS_HOSTNAME")
    
    repl_identity = os.environ.get("REPL_IDENTITY")
    web_repl_renewal = os.environ.get("WEB_REPL_RENEWAL")
    
    if repl_identity:
        x_replit_token = f"repl {repl_identity}"
    elif web_repl_renewal:
        x_replit_token = f"depl {web_repl_renewal}"
    else:
        raise Exception("X_REPLIT_TOKEN not found for repl/depl")
    
    async with httpx.AsyncClient() as client:
        response = await client.get(
            f"https://{hostname}/api/v2/connection?include_secrets=true&connector_names=resend",
            headers={
                "Accept": "application/json",
                "X_REPLIT_TOKEN": x_replit_token
            }
        )
        data = response.json()
    
    connection_settings = data.get("items", [{}])[0] if data.get("items") else None
    
    if not connection_settings or not connection_settings.get("settings", {}).get("api_key"):
        raise Exception("Resend not connected")
    
    return {
        "api_key": connection_settings["settings"]["api_key"],
        "from_email": connection_settings["settings"].get("from_email")
    }

async def send_contact_email(name: str, email: str, message: str) -> dict:
    try:
        credentials = await get_resend_credentials()
        resend.api_key = credentials["api_key"]
        
        safe_name = escape_html(name)
        safe_email = escape_html(email)
        safe_message = escape_html(message)
        
        result = resend.Emails.send({
            "from": "Backend Studio <onboarding@resend.dev>",
            "to": CONTACT_RECIPIENT,
            "subject": f"Contact Form: Message from {safe_name}",
            "html": f"""
                <h2>New Contact Form Submission</h2>
                <p><strong>From:</strong> {safe_name}</p>
                <p><strong>Email:</strong> {safe_email}</p>
                <hr />
                <h3>Message:</h3>
                <p>{safe_message.replace(chr(10), '<br>')}</p>
                <hr />
                <p style="color: #666; font-size: 12px;">
                    This message was sent from the Backend Systems Intelligence Studio contact form.
                </p>
            """,
            "text": f"""
New Contact Form Submission

From: {name}
Email: {email}

Message:
{message}

---
This message was sent from the Backend Systems Intelligence Studio contact form.
            """
        })
        
        if isinstance(result, dict) and result.get("error"):
            print(f"[Email] Resend error: {result.get('error')}")
            return {"success": False, "error": str(result.get("error"))}
        
        return {"success": True}
    
    except Exception as e:
        print(f"[Email] Email service error: {e}")
        return {"success": False, "error": str(e)}
